### PR TITLE
Make index.codec setting case-insensitive.

### PR DESCRIPTION
**Pull request by mulugetam**
### Description
OpenSearch currently supports `default`, `best_compression`, `lucene_default`, and `ZSTD` compression codecs for `index.codec`. The setting uses a case-sensitive comparison. If the user supplies `BEST_COMPRESSION`, for example, OpenSearch uses the `default` (LZ4) codec and not `best_compression` (deflate). 

This PR addresses the issue by allowing a case-insensitive match for the setting. 

### Issues Resolved
`index.codec: BEST_COMPRESSION` falls back to `default` (lz4) and not `best_compression` (deflate).

### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/engine/EngineConfig.java [hunk 1]
```diff
@@ -123,15 +124,19 @@ public Supplier<RetentionLeases> retentionLeasesSupplier() {
...
             case "best_compression":
             case "lucene_default":
                 return s;
             default:
-                if (Codec.availableCodecs().contains(s) == false) { // we don't error message the not officially supported ones
+                if (!Codec.availableCodecs().stream().anyMatch(c -> c.equalsIgnoreCase(s))) { // we don't error message the not
+                                                                                              // officially supported ones
                     throw new IllegalArgumentException(
-                        "unknown value for [index.codec] must be one of [default, best_compression] but was: " + s
+                        "unknown value for [index.codec] must be one of [default, best_compression] or "
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/engine/EngineConfig.java:136]:
```suggestion
                        "unknown value for [index.codec] must be one of [default, best_compression, lucene_default] or "
```

File: server/src/main/java/org/opensearch/index/codec/CodecService.java [hunk 2]
```diff
@@ -76,11 +77,11 @@ public CodecService(@Nullable MapperService mapperService, Logger logger) {
     }
 
     public Codec codec(String name) {
-        Codec codec = codecs.get(name);
-        if (codec == null) {
+        Optional<String> key = codecs.keySet().stream().filter(k -> k.equalsIgnoreCase(name)).findAny();
```
Comment (reta) [inline server/src/main/java/org/opensearch/index/codec/CodecService.java:80]:
```suggestion
        return codecs
            .entrySet()
            .stream()
            .filter(e -> e.getKey().equalsIgnoreCase(name))
            .map(Map.Entry::getValue)
            .findAny()
            .orElseThrow(() -> new IllegalArgumentException("failed to find codec [" + name + "]"));

```

**Comment by dblock** [comment]
> If the user supplies BEST_COMPRESSION, for example, OpenSearch uses the default (LZ4) codec and not best_compression (deflate).

I would expect an error, is this not the case? What if I specify `foobar` as a coded? Do we fallback silently?

This *is* an interface change that breaks bcw isn't it? If user was supplying `LUCENE_DEFAULT` before, were they getting `lz4` quietly and would now be getting `lucene_default`?

**Comment by mulugetam** [comment]
> > If the user supplies BEST_COMPRESSION, for example, OpenSearch uses the default (LZ4) codec and not best_compression (deflate).
> 
> I would expect an error, is this not the case? What if I specify `foobar` as a coded? Do we fallback silently?
> 
> This _is_ an interface change that breaks bcw isn't it? If user was supplying `LUCENE_DEFAULT` before, were they getting `lz4` quietly and would now be getting `lucene_default`?

My bad, apparently the silent fallback has to do with OpenSearch-Benchmark (was using it for tests) and not OpenSearch. A direct REST call to OpenSearch with `BEST_COMPRESSION` throws an exception as expected.

I think the question now is whether we should allow for a case-insensitive match. I think we should. The current compression codecs are all in lower case except ZSTD: `default`, `lucene_default`, `best_compression`, `ZSTD`. I think we should either allow a case-insensitive matching or I should change `ZSTD` to `zstd`.

**Comment by mulugetam** [comment]
OpenSearch-Benchmark bug report: https://github.com/opensearch-project/opensearch-benchmark/issues/268

@dblock @reta

**Comment by andrross** [comment]
> I think we should either allow a case-insensitive matching or I should change ZSTD to zstd.

I vote for changing `ZSTD` to `zstd` for consistency, and keeping the case-sensitive behavior. I think we can make the backward-incompatible change as ZSTD support hasn't been released yet.

**Comment by reta** [comment]
> OpenSearch-Benchmark bug report: [opensearch-project/opensearch-benchmark#268](https://github.com/opensearch-project/opensearch-benchmark/issues/268)
> 
> @dblock @reta

Thanks @mulugetam 

> I think the question now is whether we should allow for a case-insensitive match.

I would genuinely be surprised if `ZSTD` and `zstd` would be treated differently as codec name, may be we could use the case insensitive data structures for that:

```
        this.codecs = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
        this.codecs.putAll(codecs.map());
```

File: server/src/main/java/org/opensearch/index/engine/EngineConfig.java [hunk 3]
```diff
@@ -123,15 +124,19 @@ public Supplier<RetentionLeases> retentionLeasesSupplier() {
...
             case "best_compression":
             case "lucene_default":
                 return s;
             default:
-                if (Codec.availableCodecs().contains(s) == false) { // we don't error message the not officially supported ones
+                if (!Codec.availableCodecs().stream().anyMatch(c -> c.equalsIgnoreCase(s))) { // we don't error message the not
+                                                                                              // officially supported ones
                     throw new IllegalArgumentException(
-                        "unknown value for [index.codec] must be one of [default, best_compression] but was: " + s
+                        "unknown value for [index.codec] must be one of [default, best_compression, lucene_default] or "
```
Comment (Poojita-Raj) [inline server/src/main/java/org/opensearch/index/engine/EngineConfig.java:136]: If we're allowing for users to set it to lucene_default and/or one of the listed available codecs (Lucene80, Lucene70, etc), this needs to be documented on the static settings for index creation for index.codec setting. 

Ref: https://opensearch.org/docs/2.4/api-reference/index-apis/create-index/#static-index-settings

File: server/src/main/java/org/opensearch/index/engine/EngineConfig.java [hunk 4]
```diff
@@ -123,15 +124,19 @@ public Supplier<RetentionLeases> retentionLeasesSupplier() {
...
     public static final Setting<String> INDEX_CODEC_SETTING = new Setting<>("index.codec", "default", s -> {
-        switch (s) {
+        switch (s.toLowerCase(Locale.ROOT)) {
             case "default":
             case "best_compression":
             case "lucene_default":
                 return s;
             default:
-                if (Codec.availableCodecs().contains(s) == false) { // we don't error message the not officially supported ones
+                if (!Codec.availableCodecs().stream().anyMatch(c -> c.equalsIgnoreCase(s))) { // we don't error message the not
```
Comment (Poojita-Raj) [inline server/src/main/java/org/opensearch/index/engine/EngineConfig.java:133]: Nit:

```suggestion
                if (Codec.availableCodecs().stream().anyMatch(c -> c.equalsIgnoreCase(s)) == false) { // we don't error message the not
```

**Comment by reta** [comment]
> Looks like @dblock @andrross @Poojita-Raj and for a case-sensitive `index.codec`. @reta, may I close this PR? (I will create a new one that changes `ZSTD` to `zstd`)?

Sure, thanks @mulugetam !

**Comment by mulugetam** [comment]
Closing this PR and creating https://github.com/opensearch-project/OpenSearch/pull/7231 that changes the codec name from `ZSTD` to `zstd`.

@dblock @reta

**Pull request closed without merging**